### DISCUSSION TITLE
Laravel - Server Adapter

### DISCUSSION THREAD

**Discussion by alexevladgabriel** [trpc/trpc#5109] [Ideas]
tRPC is an awesome tool, it makes developer experience so much better. I enjoy using it everyday, but (there's always a but), NextJS doesn't offer some solutions for different problems like Laravel does, so if you decide to go with Laravel the type safety between these two gets lost. I would like to use Laravel with NextJS (tRPC), so my questions is:

- Would be possible to create a server adapter for Laravel? (I saw the adapters for Fastify & Express)

**Comment by sneakylenny** [comment]
I understand your interest in combining Laravel with NextJS and tRPC. However, creating a server adapter for Laravel might be challenging due to it being a PHP framework, unlike Fastify and Express.

There have been efforts to make Laravel more TypeScript-friendly, such as the spatie/typescript-transformer package (https://spatie.be/docs/typescript-transformer/v2/introduction), which requires structuring your application in a specific way.

I'm interested in exploring this idea, but I can imagine it would be a difficult challenge.

**Reply by alexevladgabriel** [reply]
Currently I'm working on a open source projects which will implement the following lib: Laravel Data.
https://spatie.be/docs/laravel-data/v4/introduction

Not sure, how the integration would be make, but can be a good start, like you said with Typescript Transformer.

**Reply by sneakylenny** [reply]
I've done it this way before, and I can say from experience that it can add some extra complexity. However, it really improves the developer experience on the front end and provides a clear, single source of truth for your data model.

I wish there were packages like Laravel-Data and Laravel-Typescript-Transformer for Symfony, as they would be much more useful there. Just keep in mind, Laravel Data is essentially an alternative to [Laravel's resources](https://laravel.com/docs/12.x/eloquent-resources), which can also be transformed using Spatie's Laravel-Typescript-Transformer.

**Reply by AlexVanderbist** [reply]
Hey, just wondering if either of you got anywhere with this idea. I work on these projects at Spatie and obviously we've been looking in the same direction for a while now. Just figuring out if trpc is a solution to this problem, or whether we want to go a more custom routes.

**Reply by sneakylenny** [reply]
### Where I am now

I have not tried to create something new yet. We currently use [Laravel JSON:RPC](https://sajya.github.io/docs/) in production, which works very well, It's a well-established package which makes use of the [JSON-RPC 2.0 specification](https://www.jsonrpc.org/specification). And if I'm not mistaken, [tRPC does this too](https://trpc.io/docs/rpc#http-response-specification). The downside is that we type our TypeScript interfaces and types in the frontend manually. So they're duplicates that we update manually, which can be very cumbersome.

### Sharing my idea

I do, however, see a potential for a package that can be built on top of Laravel JSON:RPC. It could be closely tied with Laravel Data because, the server would determine the types and DTOs are a good way to do this. And since Laravel Data works well with Laravel Typescript Transformer there's a way to translate these DTOs to TypeScript.

The missing piece of this puzzle, is automating type generation and getting these types to the frontend. 

tRPC projects are commonly monorepos. They even encourage you to use monorepos:

> It's designed for monorepos, as you need to export/import the type definitions from/to your server. 

<sup>[source](https://trpc.io/docs/further-reading#who-is-this-for)</sup>

However, it's not required:

> No, a monorepo is not mandatory but you will lose some of the benefits of using tRPC if you don't use it since you will lose guarantees that your client and server works together.

<sup>[source](https://trpc.io/docs/faq#is-a-monorepo-mandatory)</sup>

I think we can assume the majority of the projects that use tRPC or plan to use tRPC are monorepos. Keeping this in mind, it could be as simple as a script that watches relevant backend code and calls the transformer automatically. But even then, I don't know if it's possible to generate the type required (the `AppRouter` type) for a tRPC client.

### Lastly

I sadly don't have time to dive into this since I'm currently very busy working on a private project right now. I do love to discuss and read along. Let me know what you think.
